kimchi/src/tests/xor.rs

```diff
@@ -31,58 +59,145 @@ fn create_test_constraint_system(bits: usize) -> ConstraintSystem<Fp> {
     ConstraintSystem::create(gates).build().unwrap()
 }
 
-fn test_xor(in1: u128, in2: u128, bits: usize) -> [Vec<PallasField>; COLUMNS] {
-    let cs = create_test_constraint_system(bits);
-    let witness = xor::create(in1, in2, bits);
-    for row in 0..xor::num_xors(bits) + 1 {
-        assert_eq!(
-            cs.gates[row].verify_witness::<Vesta>(row, &witness, &cs, &witness[0][0..cs.public]),
-            Ok(())
-        );
+// Generates a random field element of up to a given number of bits
+pub(crate) fn random_field<G: KimchiCurve>(bits: usize, rng: &mut StdRng) -> G::ScalarField {
+    G::ScalarField::from_biguint(
+        rng.gen_biguint_range(&BigUint::from(0u8), &BigUint::from(2u8).pow(bits as u32)),
+    )
+    .unwrap()
+}
+
+// Initialize a random input
+pub(crate) fn initialize<G: KimchiCurve>(
+    input: Option<G::ScalarField>,
+    bits: Option<usize>,
+    rng: &mut StdRng,
+) -> G::ScalarField {
+    if let Some(inp) = input {
+        inp
+    } else {
+        assert!(bits.is_some());
+        let bits = bits.unwrap();
+        random_field::<G>(bits, rng)
     }
-    witness
 }
 
-// Returns a given crumb of 4 bits
-fn xor_crumb(word: u128, crumb: usize) -> u128 {
-    (word >> (4 * crumb)) % 2u128.pow(4)
+// Returns the all ones BigUint of bits length
+pub(crate) fn all_ones<G: KimchiCurve>(bits: usize) -> G::ScalarField {
+    G::ScalarField::from(2u128).pow(&[bits as u64]) - G::ScalarField::one()
 }
 
-// Manually checks the XOR of each crumb in the witness
-fn check_xor(witness: &[Vec<PallasField>; COLUMNS], bits: usize, input1: u128, input2: u128) {
-    for xor in 0..xor::num_xors(bits) {
+// Returns a given nybble of 4 bits
+pub(crate) fn xor_nybble(word: BigUint, nybble: usize) -> BigUint {
+    (word >> (4 * nybble)) % 2u128.pow(4)
+}
+
+// Manually checks the XOR of each nybble in the witness
+pub(crate) fn check_xor<G: KimchiCurve>(
```

Comment:
Why is this pub(crate) I don't code outside of tests should be using tests code.  The other way around is OK though, so maybe move this to `xor.rs`?

Reply:
Oh, but I will use this in AND, and NOT, that's why it is pub(crate) here. This is just one of the ugly parts of having so many open PRs, it is hard to get the general idea of everything when it is so partitioned...

Reply:
I see, yeah!